src/puppetlabs/puppetdb/command.clj

```diff
@@ -86,23 +103,110 @@
+   :body (s/cond-pre s/Str utils/byte-array-class)})
+
+(def new-command-schema ;; This could probably be stricter
+  {:command s/Str
+   :version s/Int
+   :certname s/Str
+   :payload {s/Any s/Any}
+   :annotations
+   (s/pred
+    (fn [x]
+      (and (map? x)
+           (some-> x :id string?)
+           (some-> x :received string?)
+           (empty? (select-keys x [:command :version :certname])))))})
+
+(def queue-command-schema
+  ;; Created to match existing test behavior, but is that
+  ;; comprehensive? i.e. if we're wrong, and this is too strict, then
+  ;; it could break on outside queue content on release.  Adding a
+  ;; schema here could change our effectively public queue API.
+  (-> new-command-schema
+      ;;(assoc :command (s/cond-pre s/Int s/Str))
+      ;;(assoc :version (s/cond-pre s/Int s/Str))
+      (dissoc :certname)
+      (assoc (s/optional-key :certname) s/Str)
+      (dissoc :payload)
+      (assoc (s/optional-key :payload) (s/cond-pre s/Str {s/Any s/Any}))))
+
+(defn-validated parse-new-command :- new-command-schema
+  "Parses a new-format queue command and returns it in the traditional
+  format (see parse-queue-command).  New-style messages must have all
+  5 headers (command, version, certname, received, and id), and the
+  body must be the bare command content (i.e. the payload).  So for a
+  \"deactivate node\" command, a string like
+  {\"certname\":\"test1\",\"producer_timestamp\":\"2015-01-01\"}."
+  [{:keys [headers body]} :- new-message-schema]
+  (let [{:keys [command version certname received id]} headers
+        version (some-> headers :version Integer/parseInt)]
+    (let [message (json/parse body true)]
+      (die-on-header-payload-mismatch "certname"
+                                      certname
+                                      (get-in message ["payload" "certname"]))
+      ;; Since new commands aren't the queue retry format, we expect
+      ;; no annotations.
+      (assert (not (:annotations message)))
+      {:command command
+       :version version
+       :certname certname ;; Duplicated with respect to payload for now.
+       :payload message
+       :annotations (dissoc headers :command :version :certname)})))
+
+(defn-validated parse-queue-command :- queue-command-schema
+  "Parses a traditional queue command (also the current format for
+  failed, re-queued messages).  See parse-new-command for the handling
+  of newly received messages.  These messages must not have command,
+  version, or certname headers, and the body must be a JSON map like
+  this: {\"command\":\"deactivate node\",\"version\":3,\"payload\":{\"certname\":...}}."
+  [{:keys [headers body]} :- queue-message-schema]
+  (let [{:keys [command version certname]} headers]
+    (update (json/parse-strict body true) :annotations merge
+            ;; cf. PDB-2315
+            {:received (kitchensink/timestamp) :id (kitchensink/uuid)}
+            headers)))
+
+(defn-validated parse-command
```

Comment:
If I understand you, that's intentional since the validation's already been handled by the two old/new sub-calls, and I didn't want to duplicate the work.